File: IOS/AttendanceCore/Sources/AttendanceCore/MaitriClient.swift
import Foundation

public final class MaitriClient: Sendable {
    public init() {}

    public func fetchAttendance(creds: Credentials) async throws -> AttendanceSnapshot {
        guard creds.username.contains("@") else { throw MaitriError.usernameNotEmail }
        guard !creds.password.isEmpty else { throw MaitriError.invalidCredentials }

        let cookieStorage = EphemeralCookieStorage()
        let session = Self.makeSession()
        let transport = MaitriTransport(session: session, cookieStorage: cookieStorage)

        do {
            try await transport.seedSession()
            try await transport.authenticate(creds: creds)
            let terms = try await transport.fetchTerms()
            guard let current = terms.max(by: { $0.semesterId < $1.semesterId }) else {
                throw MaitriError.noTerms
            }
            let subjects = try await transport.fetchSubjects(term: current)
            return AttendanceSnapshot(
                termName: current.name,
                termSemesterId: current.semesterId,
                subjects: subjects,
                fetchedAtMillis: Int64(Date().timeIntervalSince1970 * 1000)
            )
        } catch let error as MaitriError {
            throw error
        } catch let error as URLError {
            throw MaitriError.network(error.localizedDescription)
        } catch {
            throw MaitriError.portal(error.localizedDescription)
        }
    }

    private static func makeSession() -> URLSession {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 20
        configuration.timeoutIntervalForResource = 35
        configuration.httpShouldSetCookies = true
        configuration.httpCookieAcceptPolicy = .always
        return URLSession(configuration: configuration)
    }
}

private final class EphemeralCookieStorage: @unchecked Sendable {
    private let lock = NSLock()
    private var cookies: [HTTPCookie] = []

    func setCookies(_ incoming: [HTTPCookie]) {
        lock.lock()
        defer { lock.unlock() }
        let names = Set(incoming.map(\.name))
        cookies.removeAll { names.contains($0.name) }
        cookies.append(contentsOf: incoming)
    }

    func cookies(for url: URL) -> [HTTPCookie] {
        lock.lock()
        defer { lock.unlock() }
        return cookies.filter { cookie in
            guard let host = url.host else { return false }
            return host.hasSuffix(cookie.domain.trimmingCharacters(in: CharacterSet(charactersIn: ".")))
        }
    }
}

private struct MaitriTransport {
    private static let base = URL(string: "https://maitri.bmu.edu.in")!
    private static let userAgent =
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 " +
        "(KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"

    let session: URLSession
    let cookieStorage: EphemeralCookieStorage

    func seedSession() async throws {
        let url = Self.base.appendingPathComponent("loginPage.htm")
        _ = try await data(for: URLRequest(url: url), acceptHTML: true)
    }

    func authenticate(creds: Credentials) async throws {
        var request = URLRequest(url: Self.base.appendingPathComponent("j_spring_security_check"))
        request.httpMethod = "POST"
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("\(Self.base.absoluteString)/login.htm", forHTTPHeaderField: "Referer")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let body = "j_username=\(formEncode(creds.username))&j_password=\(formEncode(creds.password))"
        request.httpBody = body.data(using: .utf8)

        let (_, response) = try await data(for: request, acceptHTML: true)
        let finalURL = response.url?.absoluteString ?? ""
        if finalURL.localizedCaseInsensitiveContains("login.htm") {
            throw MaitriError.invalidCredentials
        }
    }

    func fetchTerms() async throws -> [Term] {
        let url = Self.base.appendingPathComponent("stu_getTermsOfStudentForCourceFile.json")
        let data = try await jsonData(for: url)
        let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] ?? []
        return array.compactMap { object in
            guard let semesterId = object["semesterId"] as? Int else { return nil }
            return Term(
                semesterId: semesterId,
                name: object["terms2Name"] as? String ?? "Sem \(semesterId)",
                startDate: object["teachingStartDate"] as? String ?? "?",
                endDate: object["teachingEndDate"] as? String ?? "?",
                subjectwiseStudentIds: object["subjectwiseStudentIds"] as? String ?? "",
                subejctwiseBatchIds: object["subejctwiseBatchIds"] as? String ?? "",
                batchsemesterCapacity: object["batchsemesterCapacity"] as? Int ?? 0
            )
        }
    }

    func fetchSubjects(term: Term) async throws -> [Subject] {
        var components = URLComponents(
            url: Self.base.appendingPathComponent("stu_getSubjectOnChangeWithSemId1.json"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "termId", value: String(term.semesterId)),
            URLQueryItem(name: "refreshData", value: "0"),
            URLQueryItem(name: "subjectwisestudentids", value: term.subjectwiseStudentIds),
            URLQueryItem(name: "subejctwiseBatchIds", value: term.subejctwiseBatchIds),
            URLQueryItem(name: "batchsemestercapacity", value: String(term.batchsemesterCapacity)),
        ]
        let data = try await jsonData(for: components.url!)
        let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] ?? []
        return array.compactMap { object in
            if let category = object["subjectCategory"] as? String, category == "PROJECT" {
                return nil
            }
            return Subject(
                code: object["subjectCode"] as? String ?? "?",
                name: decodeHtmlEntities(object["subject"] as? String ?? "?"),
                present: object["presentCount"] as? Int ?? 0,
                absent: object["absentCount"] as? Int ?? 0,
                afterCapping: object["afterCapping"] as? Double ?? 0
            )
        }
    }

    private func jsonData(for url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json, text/plain, */*", forHTTPHeaderField: "Accept")
        request.setValue("XMLHttpRequest", forHTTPHeaderField: "X-Requested-With")
        let (data, response) = try await data(for: request, acceptHTML: false)
        let finalURL = response.url?.absoluteString ?? ""
        if finalURL.localizedCaseInsensitiveContains("login.htm") {
            throw MaitriError.portal("Session was rejected when calling \(url.path)")
        }
        guard !data.isEmpty else {
            throw MaitriError.portal("Empty response body from \(url.path)")
        }
        return data
    }

    private func data(for request: URLRequest, acceptHTML: Bool) async throws -> (Data, HTTPURLResponse) {
        var request = request
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        if acceptHTML {
            request.setValue("text/html,application/json,*/*", forHTTPHeaderField: "Accept")
        }
        if let url = request.url {
            let cookies = cookieStorage.cookies(for: url)
            if !cookies.isEmpty {
                let header = cookies.map { "\($0.name)=\($0.value)" }.joined(separator: "; ")
                request.setValue(header, forHTTPHeaderField: "Cookie")
            }
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw MaitriError.portal("Unexpected response type")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw MaitriError.portal("HTTP \(http.statusCode) for \(request.url?.path ?? "?")")
        }
        if let url = request.url, let headerFields = http.allHeaderFields as? [String: String] {
            let parsed = HTTPCookie.cookies(withResponseHeaderFields: headerFields, for: url)
            if !parsed.isEmpty {
                cookieStorage.setCookies(parsed)
            }
        }
        return (data, http)
    }

    private func formEncode(_ value: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    private func decodeHtmlEntities(_ value: String) -> String {
        value
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: "&#039;", with: "'")
            .replacingOccurrences(of: "&nbsp;", with: " ")
    }
}

File: Android/app/app/src/main/kotlin/edu/bmu/attendance/data/MaitriClient.kt
package edu.bmu.attendance.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.FormBody
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * The single source of truth for talking to Maitri.
 *
 * Performs the login → terms → attendance flow described in `Android/FLOW.md`
 * and returns a fully-populated [AttendanceSnapshot]. Each call is
 * self-contained: it logs in fresh, fetches, and lets the cookie jar GC. We
 * deliberately do not persist the `JSESSIONID` across calls — that lets us
 * recover from "session expired" failures simply by calling [fetchAttendance]
 * again, with no tricky retry logic.
 *
 * All public functions are `suspend` and dispatch network I/O on
 * [Dispatchers.IO].
 */
class MaitriClient(
    private val httpClient: OkHttpClient = defaultHttpClient(),
) {
    /**
     * Logs in and pulls the latest [AttendanceSnapshot] for the term with the
     * highest `semesterId` (i.e. the most recently enrolled).
     *
     * Throws a [MaitriError] subclass on any expected failure; rethrows
     * unexpected exceptions as [MaitriError.Network] / [MaitriError.Portal].
     */
    suspend fun fetchAttendance(creds: Credentials): AttendanceSnapshot =
        withContext(Dispatchers.IO) {
            withAuthenticatedClient(creds) { client, current ->
                buildAttendanceSnapshot(client, current)
            }
        }

    /**
     * Logs in and pulls the student timetable for [startDate]…[endDate]
     * (portal labels like `Aug 2, 2026`). Defaults to the current Sun–Sat week.
     */
    suspend fun fetchTimetable(
        creds: Credentials,
        startDate: String? = null,
        endDate: String? = null,
    ): TimetableSnapshot =
        withContext(Dispatchers.IO) {
            val range = TimetableDateRange.currentWeek(startLabel = startDate, endLabel = endDate)
            withAuthenticatedClient(creds) { client, current ->
                buildTimetableSnapshot(client, current, range)
            }
        }

    /**
     * One login, then attendance + current-week timetable. Timetable failure
     * does not fail the whole call — attendance still returns and timetable
     * is null so callers can keep the previous cached week.
     */
    suspend fun fetchAttendanceAndTimetable(creds: Credentials): Pair<AttendanceSnapshot, TimetableSnapshot?> =
        withContext(Dispatchers.IO) {
            val range = TimetableDateRange.currentWeek()
            withAuthenticatedClient(creds) { client, current ->
                val attendance = buildAttendanceSnapshot(client, current)
                val timetable = runCatching {
                    buildTimetableSnapshot(client, current, range)
                }.getOrNull()
                attendance to timetable
            }
        }

    private fun buildAttendanceSnapshot(client: OkHttpClient, current: Term): AttendanceSnapshot {
        val subjects = fetchSubjects(client, current)
        return AttendanceSnapshot(
            termName = current.name,
            termSemesterId = current.semesterId,
            subjects = subjects,
            fetchedAtMillis = System.currentTimeMillis(),
        )
    }

    private fun buildTimetableSnapshot(
        client: OkHttpClient,
        current: Term,
        range: TimetableDateRange,
    ): TimetableSnapshot {
        val sessions = fetchTimetableSessions(client, current, range.startLabel, range.endLabel)
        return TimetableSnapshot(
            termName = current.name,
            termSemesterId = current.semesterId,
            startDate = range.startLabel,
            endDate = range.endLabel,
            sessions = sessions,
            fetchedAtMillis = System.currentTimeMillis(),
        )
    }

    private fun <T> withAuthenticatedClient(
        creds: Credentials,
        work: (OkHttpClient, Term) -> T,
    ): T {
        if (!creds.username.contains('@')) throw MaitriError.UsernameNotEmail
        if (creds.password.isBlank()) throw MaitriError.InvalidCredentials

        // Each call uses its own ephemeral cookie jar.
        val callJar = InMemoryCookieJar()
        val client = httpClient.newBuilder().cookieJar(callJar).build()

        try {
            seedSession(client)
            authenticate(client, creds)
            val terms = fetchTerms(client)
            val current = terms.maxByOrNull { it.semesterId } ?: throw MaitriError.NoTerms
            return work(client, current)
        } catch (e: MaitriError) {
            throw e
        } catch (e: IOException) {
            throw MaitriError.Network(e)
        } catch (e: org.json.JSONException) {
            throw MaitriError.Portal(e.message ?: "Portal returned invalid JSON")
        } catch (e: IllegalStateException) {
            throw MaitriError.Portal(e.message ?: "Portal returned unexpected data")
        } catch (e: Exception) {
            throw MaitriError.Portal(e.message ?: e::class.java.simpleName)
        }
    }

    // ── Internal steps ──────────────────────────────────────────────────────

    private fun seedSession(client: OkHttpClient) {
        client.newCall(get(LOGIN_PAGE_URL)).execute().use { resp ->
            check(resp.isSuccessful) { "Login page returned HTTP ${resp.code}" }
        }
    }

    private fun authenticate(client: OkHttpClient, creds: Credentials) {
        val body = FormBody.Builder()
            .add("j_username", creds.username)
            .add("j_password", creds.password)
            .build()
        val request = Request.Builder()
            .url(LOGIN_CHECK_URL)
            .header("User-Agent", USER_AGENT)
            .header("Referer", "$BASE/login.htm")
            .post(body)
            .build()
        client.newCall(request).execute().use { resp ->
            val finalUrl = resp.request.url.toString()
            check(resp.isSuccessful) {
                "Login POST returned HTTP ${resp.code} (final URL: $finalUrl)"
            }
            if (finalUrl.contains("login.htm", ignoreCase = true)) {
                throw MaitriError.InvalidCredentials
            }
        }
    }

    private fun fetchTerms(client: OkHttpClient): List<Term> {
        val json = fetchJson(client, TERMS_URL.toHttpUrl())
        val arr = JSONArray(json)
        return List(arr.length()) { i ->
            val o = arr.getJSONObject(i)
            Term(
                semesterId = o.getInt("semesterId"),
                name = o.optString("terms2Name", "Sem ${o.getInt("semesterId")}"),
                startDate = o.optString("teachingStartDate", "?"),
                endDate = o.optString("teachingEndDate", "?"),
                subjectwiseStudentIds = o.optString("subjectwiseStudentIds", ""),
                subejctwiseBatchIds = o.optString("subejctwiseBatchIds", ""),
                batchsemesterCapacity = o.optInt("batchsemesterCapacity", 0),
            )
        }
    }

    private fun fetchSubjects(client: OkHttpClient, term: Term): List<Subject> {
        val url = SUBJECTS_URL.toHttpUrl().newBuilder().apply {
            addQueryParameter("termId", term.semesterId.toString())
            addQueryParameter("refreshData", "0")
            addQueryParameter("subjectwisestudentids", term.subjectwiseStudentIds)
            addQueryParameter("subejctwiseBatchIds", term.subejctwiseBatchIds)
            addQueryParameter("batchsemestercapacity", term.batchsemesterCapacity.toString())
        }.build()
        val json = fetchJson(client, url)
        val arr = JSONArray(json)
        return buildList {
            for (i in 0 until arr.length()) {
                val o = arr.getJSONObject(i)
                if (o.optString("subjectCategory") == "PROJECT") continue
                add(
                    Subject(
                        code = o.optString("subjectCode", "?"),
                        name = decodeHtmlEntities(o.optString("subject", "?")),
                        present = o.optInt("presentCount", 0),
                        absent = o.optInt("absentCount", 0),
                        afterCapping = o.optDouble("afterCapping", 0.0),
                    ),
                )
            }
        }
    }

    private fun fetchTimetableSessions(
        client: OkHttpClient,
        term: Term,
        startDate: String,
        endDate: String,
    ): List<TimetableSession> {
        val url = TIMETABLE_URL.toHttpUrl().newBuilder().apply {
            addQueryParameter("startDate", startDate)
            addQueryParameter("endDate", endDate)
            addQueryParameter("termId", term.semesterId.toString())
        }.build()
        val json = fetchJson(client, url)
        val arr = JSONArray(json)
        return List(arr.length()) { i ->
            val o = arr.getJSONObject(i)
            TimetableSession(
                lectureDate = o.optString("lectureDate", "").trim(),
                dateIso = o.optString("lectureDateWithoutFormat", ""),
                lectureDay = o.optString("lectureDay", ""),
                startTime = o.optString("lectureStartTime", ""),
                endTime = o.optString("lectureEndTime", ""),
                // Upstream typo — keep reading `sessoionNo`.
                sessionNo = o.opt("sessoionNo")?.toString()?.takeIf { it != "null" }.orEmpty(),
                subjectName = decodeHtmlEntities(o.optString("subjectName", "")),
                topic = decodeUriComponent(o.optString("chapterName", "-")),
                classRoom = o.optString("classRoom", "-"),
                facultyName = collapseWhitespace(o.optString("facultyName", "")),
                description = o.optString("guestLecDescription", "-"),
            )
        }
    }

    private fun collapseWhitespace(value: String): String =
        value.split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ")

    private fun decodeUriComponent(value: String): String {
        var current = value
        repeat(2) {
            val decoded = runCatching {
                java.net.URLDecoder.decode(current, Charsets.UTF_8.name())
            }.getOrDefault(current)
            if (decoded == current) return@repeat
            current = decoded
        }
        val normalized = decodeHtmlEntities(current)
        return normalized.ifBlank { "-" }
    }

    private fun fetchJson(client: OkHttpClient, url: HttpUrl): String {
        val request = Request.Builder()
            .url(url)
            .header("User-Agent", USER_AGENT)
            .header("Accept", "application/json, text/plain, */*")
            .header("X-Requested-With", "XMLHttpRequest")
            .build()
        return client.newCall(request).execute().use { resp ->
            check(resp.isSuccessful) { "GET ${url.encodedPath} returned HTTP ${resp.code}" }
            val finalUrl = resp.request.url.toString()
            check(!finalUrl.contains("login.htm", ignoreCase = true)) {
                "Session was rejected when calling ${url.encodedPath}"
            }
            val body = resp.body!!.string()
            check(body.isNotBlank()) { "Empty response body from ${url.encodedPath}" }
            body
        }
    }

    private fun get(url: String): Request = Request.Builder()
        .url(url)
        .header("User-Agent", USER_AGENT)
        .header("Accept", "text/html,application/json,*/*")
        .build()

    /**
     * Maitri's `subject` field passes course names through a `gemsDecode()`
     * helper in browser-side JS. The bytes that arrive over the wire usually
     * already look fine, but we also normalise common HTML entities just in
     * case (e.g. `&amp;`, `&#039;`).
     */
    private fun decodeHtmlEntities(s: String): String =
        s.replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            .replace("&#039;", "'")
            .replace("&nbsp;", " ")

    companion object {
        private const val BASE = "https://maitri.bmu.edu.in"
        private const val LOGIN_PAGE_URL = "$BASE/loginPage.htm"
        private const val LOGIN_CHECK_URL = "$BASE/j_spring_security_check"
        private const val TERMS_URL = "$BASE/stu_getTermsOfStudentForCourceFile.json"
        private const val SUBJECTS_URL = "$BASE/stu_getSubjectOnChangeWithSemId1.json"
        private const val TIMETABLE_URL = "$BASE/stu_getBetweenDatesTimetableForStudentSP.json"
        private const val USER_AGENT =
            "Mozilla/5.0 (Linux; Android 14; Pixel 8) AppleWebKit/537.36 " +
                "(KHTML, like Gecko) Chrome/129.0.0.0 Mobile Safari/537.36"

        fun defaultHttpClient(): OkHttpClient = OkHttpClient.Builder()
            .followRedirects(true)
            .followSslRedirects(true)
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.SECONDS)
            .callTimeout(35, TimeUnit.SECONDS)
            .build()
    }
}

/** Bare-bones in-memory CookieJar — one instance per `fetchAttendance` call. */
private class InMemoryCookieJar : CookieJar {
    private val store = mutableListOf<Cookie>()

    @Synchronized
    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        val incomingNames = cookies.map { it.name }.toSet()
        store.removeAll { it.name in incomingNames }
        store.addAll(cookies)
    }

    @Synchronized
    override fun loadForRequest(url: HttpUrl): List<Cookie> =
        store.filter { it.matches(url) }
}
